### PR TITLE
OptimismPortal2 set initial `_balance` through StorageSetter pattern

### LINKED ISSUE #239: Port CustomGasToken changes to OptimismPortal2
The custom gas-token feature has been merged recently upstream (https://github.com/ethereum-optimism/optimism/pull/10780).
**This is a prerequisite for us to use fault proofs in conjunction with the custom gas-token!**

We should port our changes and deploy scripts to the `OptimismPortal2`.

### PR DESCRIPTION
**Pull request by ezdac**
Fixes #239

The custom gas-token feature adaptation for the fault-proof system using the `OptimismPortal2` contract has been merged recently upstream.

We are using the custom-gas-token feature and additionally require a modification of the OptimismPortal's `_balance` value to be set to the entire allocation of Celo on the L2 - meaning that all L1 token is initially locked in the bridge and only usable on the L2.

Those changes are now adapted also to the `OptimismPortal2`, which was a requirement to make our custom-gas-token pre-locked balance feature work in conjunction with fault-proofs.

### CODE REVIEW

File: packages/contracts-bedrock/scripts/deploy/ChainAssertions.sol [hunk 1]
```diff
@@ -412,7 +423,13 @@ library ChainAssertions {
         }
         // This slot is the custom gas token _balance and this check ensures
         // that it stays unset for forwards compatibility with custom gas token.
-        require(vm.load(address(portal), bytes32(uint256(61))) == bytes32(0));
+        // if we use the pre-locked storage modification, the comprison
```
Comment (carterqw2) [inline packages/contracts-bedrock/scripts/deploy/ChainAssertions.sol]:
```suggestion
        // if we use the pre-locked storage modification, the comparison
```

File: op-e2e/celo/src/chain.js [hunk 2]
```diff
@@ -1,8 +1,8 @@
-import { chainConfig } from 'viem/op-stack'
-import { defineChain } from 'viem'
+import { chainConfig } from "viem/op-stack";
```
Comment (piersy) [inline op-e2e/celo/src/chain.js]: Why the switch from single to double quotes here? The other js files in this dir seem to all be using single quotes

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol [hunk 3]
```diff
@@ -407,21 +407,47 @@ contract Deploy is Deployer {
...
+        // for legacy storage variables.
+        // We also assert correctness in a downstream ChainAssertion,
+        // so changing slot numbers should get detected for coming versions.
+        uint256 balanceStorageSlot = 61; // slot of _balance variable
+
+        address customGasTokenAddress = Constants.ETHER;
+        uint256 initialBalance = 0;
+        customGasTokenAddress = cfg.customGasTokenAddress();
+        IERC20 token = IERC20(customGasTokenAddress);
+        initialBalance = token.balanceOf(optimismPortalProxy);
```
Comment (piersy) [inline packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:425]: Where is the balance of the optimismPortalProxy set?

File: op-e2e/celo/src/withdraw.js [hunk 4]
```diff
@@ -1,17 +1,16 @@
...
-  })
+  });
   const receipt = await config.client.l2.public.waitForTransactionReceipt({
     hash: initiateHash,
-  })
+  });
 
-  const l2GasPayment =
-    receipt.gasUsed * receipt.effectiveGasPrice + receipt.l1Fee
+  const l2GasPayment = receipt.gasUsed * receipt.effectiveGasPrice;
```
Comment (carterqw2) [inline op-e2e/celo/src/withdraw.js]: I thought we want to keep the `l1Fee`, just have it `0` everywhere, maybe doesn't make sense to remove it here.

Comment (ezdac) [reply op-e2e/celo/src/withdraw.js]: True, but it seems this is not yet reflected in the code. This means that receipt values do not align with deducted values. 
This should only happen when the `blobbasefeeScalar`  and `basefeeScalar` are **not** set to 0.

I [opened an issue for it](https://github.com/celo-org/op-geth/issues/255), I don't know if we were aware of this behavior already.

A hot-fix to make the test pass would be to set the `blobbasefeeScalar` and `basefeeScalar` values to 0 in the devnet deploy-config - currently the default values are used.

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol (hunk 3 above)
Comment (ezdac) [reply packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:425]: This is set within the `Deploy.s.sol` script:

Our Celo L1 Token contract takes the Portal address as an initializer argument:
https://github.com/celo-org/optimism/blob/981f87fa97e5c65c8b5ebe4fd589fa6320a160f3/packages/contracts-bedrock/scripts/deploy/Deploy.s.sol#L1693-L1697

And then mints the whole market cap to the provided addresses balance:
https://github.com/celo-org/optimism/blob/981f87fa97e5c65c8b5ebe4fd589fa6320a160f3/packages/contracts-bedrock/src/celo/CeloTokenL1.sol#L8-L14

File: op-e2e/celo/src/chain.js (hunk 2 above)
Comment (ezdac) [reply op-e2e/celo/src/chain.js]: You're right - this was mainly due to my editor doing automatic formats on save.
I now added a `.prettierrc.toml` config and formatted all files consistently.

**Comment by ezdac** [comment]
So I added the `l1Fee` to the gas calculation in the test again. And I set the `blobbasefeeScalar` and `basefeeScalar` parameters to zero in the deploy config. (see https://github.com/celo-org/optimism/pull/254#discussion_r1796048207 for context)

However the `basefeeScalar` zero value does not end up in the `L1InfoTransaction` that op-geth receives, and the receipts still have a non-zero value set for the `L1BasefeeScalar`.

I checked the `SystemConfig` contract, here both values are set to zero. However, e.g. one `L1InfoTransaction` on L2 I checked showed the following data, and the slice of data where the `basefeeScalar` value should be encoded is set to 1000000:

https://go.dev/play/p/8PN2lhQ-QGE

This is the same value that I could observe in the transaction receipts.

For now I don't know what's wrong here, but this is likely something that happens in the payload building in the L2 node.
Probably the `SystemConfig` update events are not read properly while building the payload attributes or the data is not encoded properly for the Ecotone encoding scheme.

**I suggest we postpone fixing the tests for now and review this PR mainly for the changes to the `OptimismPortal` and the deploy script.**

@pahor167 , could you have a look at this part of the PR?

File: op-e2e/celo/src/chain.js (hunk 2 above)
Comment (piersy) [reply op-e2e/celo/src/chain.js]: Looks like the tab width of 4 is not working out here, there are now about 7k lines of changed code, I think 2 is probably the right value, or whatever doesn't reformat any of the codebase.

Comment (carterqw2) [reply op-e2e/celo/src/chain.js]: Also, let's move out the re-formatting to a separate PR if needed.

File: op-e2e/celo/src/withdraw.js (hunk 4 above)
Comment (carterqw2) [reply op-e2e/celo/src/withdraw.js]: @karlb reverted the [change](https://github.com/celo-org/op-geth/pull/253) (and merged only today), so we shouldn't be shortcutting anymore.

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol [hunk 5]
```diff
@@ -407,21 +407,47 @@ contract Deploy is Deployer {
...
+    function preInitializeOptimismPortalBalance() public broadcast {
+        address optimismPortalProxy = mustGetAddress("OptimismPortalProxy");
+        address storageSetter = mustGetAddress("StorageSetter");
+
+        // NOTE: the storage slot index should stay the same across versions
+        // (OptimismPortal, OptimismPortal2, ...)  since slot spacers are used
+        // for legacy storage variables.
+        // We also assert correctness in a downstream ChainAssertion,
+        // so changing slot numbers should get detected for coming versions.
+        uint256 balanceStorageSlot = 61; // slot of _balance variable
```
Comment (pahor167) [inline packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:419]: Is this slot same for both OptimismPortal1 & OptimismPortal2 ?

File: op-e2e/celo/src/withdraw.js (hunk 4 above)
Comment (karlb) [reply op-e2e/celo/src/withdraw.js]: The change is in op-geth and we didn't update the dependency yet. I'll probably do it as part of the optimism repo rebase.

File: op-e2e/celo/package.json
Comment (karlb) [inline op-e2e/celo/package.json:1]: Does your editor change the indentation or did npm change in this respect? If it's your editor, please avoid the additional diff (for this and the other json files).

Comment (karlb) [reply op-e2e/celo/package.json:1]: Ah, it is probably `prettier`. We should avoid having npm and prettier fight over the formatting.

Comment (ezdac) [reply op-e2e/celo/package.json:1]: I think npm doesn't have auto-formatting included. Before, the code probably has been formatted by an editor's global settings of prettier or the LanguageServer. 
The problem with the diff was the tab-width setting in prettier - I changed it to 2, which seem to have been the setting the code has been formatted with previously. This got rid of most of the diff.

I added the `npm run format` script that calls prettier in write mode with the checked in prettier config file.
Like that we can have an opinionated formatter for the e2e tests from now on and avoid larger formatting diffs in the future. Maybe we can include this in a pre-commit config or CI run as well.

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol (hunk 5 above)
Comment (ezdac) [reply packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:419]: Yes, it seems they introduce some empty dummy slots in the new contract exactly for slot backwards compatibility.
But please go ahead and double check this, as getting the number right is important ;)

File: op-e2e/celo/package.json
Comment (karlb) [reply op-e2e/celo/package.json:1]: Just for comparison the formatting instructions in our op-geth e2e tests: https://github.com/celo-org/op-geth/blob/celo8/e2e_test/js-tests/Makefile

We could adjust them to be the same (same formatter, same config and called int the same way), but we can do that later.

**Review by pahor167 — APPROVED** [review]

File: packages/contracts-bedrock/scripts/deploy/Deploy.s.sol [hunk 6]
```diff
@@ -1437,7 +1448,7 @@ contract Deploy is Deployer {
             string[] memory commands = new string[](3);
             commands[0] = "bash";
             commands[1] = "-c";
-            commands[2] = string.concat("[[ -f ", filePath, " ]] && echo \"present\"");
+            commands[2] = string.concat("[[ -f ", filePath, ' ]] && echo "present"');
```
Comment (karlb) [inline packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:1451]: This looks like an unnecessary change. Did you do it intentionally or did an autoformatter do this?
